### PR TITLE
[MLIR][XLA] Add ConcatenateOp to LHLO/HLO emitters

### PR DESCRIPTION
**Pull request by xinan-jiang**
This is a PR from JIZHI, the AI platform in Tencent.
@sherhut @pifon2a 

We work on TensorFlow/MLIR to make mlir_gpu enable.

### CODE REVIEW

File: tensorflow/compiler/xla/service/mlir_gpu/lhlo_dialect_emitter.cc [hunk 1]
```diff
@@ -254,6 +255,20 @@ Status LhloDialectEmitter::HandleBroadcast(HloInstruction* instr) {
...
+  mlir::IntegerAttr concatenate_dim = builder_.getI64IntegerAttr(
+      static_cast<HloConcatenateInstruction*>(instr)->concatenate_dimension());
+
+  TF_ASSIGN_OR_RETURN(auto function, CreateFunction(*instr));
+  OpBuilder func_builder(function.getBody());
+  llvm::SmallVector<Value, 4> arg_values{function.args_begin(),
+                                         function.args_end()};
+  Value result_memref = function.getArgument(function.getNumArguments() - 1);
+  func_builder.create<lhlo::ConcatenateOp>(
+      getLocation(instr), arg_values, result_memref, concatenate_dim);
```
Comment (sherhut) [inline tensorflow/compiler/xla/service/mlir_gpu/lhlo_dialect_emitter.cc]: This seems to pass the result memref twice. How about `function.getArguments().drop_back(1)` for the arguments and `function.getArguments().back()`? These (or similar) should exist.

File: tensorflow/compiler/xla/service/mlir_gpu/lhlo_dialect_emitter.h [hunk 2]
```diff
@@ -17,6 +17,7 @@ limitations under the License.
 #define TENSORFLOW_COMPILER_XLA_SERVICE_MLIR_GPU_LHLO_DIALECT_EMITTER_H_
 
 #include <memory>
+#include <utility>
```
Comment (sherhut) [inline tensorflow/compiler/xla/service/mlir_gpu/lhlo_dialect_emitter.h]: What is this needed for?

File: tensorflow/compiler/xla/service/mlir_gpu/hlo_dialect_emitter.cc [hunk 3]
```diff
@@ -134,6 +135,22 @@ Status HloDialectEmitter::HandleBroadcast(HloInstruction* instr) {
   return Status::OK();
 }
 
+Status HloDialectEmitter::HandleConcatenate(HloInstruction* instr) {
+  int64 concatenate_dim = instr->concatenate_dimension();
+  TF_ASSIGN_OR_RETURN(Type res_type, ConvertTensorShapeToType<RankedTensorType>(
+                                         instr->shape(), builder_));
+
+  llvm::SmallVector<Value, 4> arguments;
```
Comment (sherhut) [inline tensorflow/compiler/xla/service/mlir_gpu/hlo_dialect_emitter.cc]: Can you add `arguments.reserve(...)`?

**Review by sherhut — CHANGES_REQUESTED** [review]
Thanks. Could you add a test?

File: tensorflow/compiler/xla/service/mlir_gpu/lhlo_dialect_emitter.h (hunk 2 above)
Comment (xinan-jiang) [reply tensorflow/compiler/xla/service/mlir_gpu/lhlo_dialect_emitter.h]: line73: std::move

File: tensorflow/compiler/xla/service/mlir_gpu/lhlo_dialect_emitter.cc (hunk 1 above)
Comment (xinan-jiang) [reply tensorflow/compiler/xla/service/mlir_gpu/lhlo_dialect_emitter.cc]: Thanks, I made a mistake here.

File: tensorflow/compiler/xla/service/mlir_gpu/hlo_dialect_emitter.cc (hunk 3 above)
Comment (xinan-jiang) [reply tensorflow/compiler/xla/service/mlir_gpu/hlo_dialect_emitter.cc]: Good suggestion. Done

**Comment by xinan-jiang** [comment]
> Thanks. Could you add a test?

The test is added. Please help to review it. Thanks.

File: tensorflow/compiler/xla/service/mlir_gpu/lhlo_dialect_emitter.h (hunk 2 above)
Comment (sherhut) [reply tensorflow/compiler/xla/service/mlir_gpu/lhlo_dialect_emitter.h]: Thanks, I did not see a change.

**Review by sherhut — APPROVED** [review]
Thanks for adding this!

**Comment by xinan-jiang** [comment]
Do I need to squash the two commits?

**Comment by sherhut** [comment]
> Do I need to squash the two commits?

That should not be needed.

**Pull request merged**